src/main/java/com/limechain/trie/DiskTrieService.java

```diff
@@ -0,0 +1,826 @@
+            Optional<Nibbles> result = findNextKey(prefix, nextPath, childNode);
+            if (result.isPresent()) {
+                return result;
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    public void upsertNode(byte[] merkleRoot, Nibbles key, byte[] storageValue, StateVersion stateVersion) {
+        TreeMap<Nibbles, PendingTrieNodeChange> executionUpdates = new TreeMap<>();
+
+        TraversalResult traversalResult = traverseTrie(merkleRoot, key);
+        switch (traversalResult) {
+            case TraversalResult.NotFound notFound -> executionUpdates.putAll(
+                executeInsert(
+                    merkleRoot,
+                    key,
+                    storageValue,
+                    stateVersion,
+                    notFound.getClosestAncestor().orElse(null)));
+            case TraversalResult.Found found -> {
+                TraversedNode foundNode = found.getFound();
+
+                assert key.equals(foundNode.getFullKey())
+                    : "Found node in traversal result should have the same key path as the sought key";
+
+                executionUpdates.put(foundNode.getFullKey(), toPendingInsertUpdate(
+                    storageValue,
+                    foundNode.getPartialKey(),
+                    foundNode.getStateVersion(),
+                    foundNode.getChildrenMerkleValues(),
+                    foundNode.getParent() == null
+                ));
+            }
+            case TraversalResult.Unfinished ignored ->
+                throw new IllegalStateException("Traversal result cannot be unfinished at this point in the logic");
+        }
+
+        Map.Entry<Nibbles, PendingTrieNodeChange> closestSuccessor = executionUpdates.firstEntry();
+        executionUpdates.putAll(toPendingInsertUpdates(traversalResult.traversedNodes,
+            closestSuccessor.getKey(),
+            closestSuccessor.getValue()));
+
+        trieChanges.updateCache(executionUpdates);
+    }
+
+    private TreeMap<Nibbles, PendingInsertUpdate> executeInsert(byte[] merkleRoot,
+                                                                Nibbles key,
+                                                                byte[] newNodeValue,
+                                                                StateVersion newNodeStateVersion,
+                                                                @Nullable TraversedNode ancestor) {
+        TreeMap<Nibbles, PendingInsertUpdate> result = new TreeMap<>();
+
+        TrieNodeData rootNode = trieStorage.getTrieNodeFromMerkleValue(merkleRoot);
+        List<byte[]> newNodeChildren = new ArrayList<>(Collections.nCopies(16, null));
+
+        if (rootNode == null) {
+            result.put(key, toPendingInsertUpdate(
+                newNodeValue,
+                key,
+                newNodeStateVersion,
+                newNodeChildren,
+                true));
+
+            return result;
+        }
+
```

Comment:
you can segregate this private function into several, it's way too long currently.